components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,229 @@
+    return;
+  }
+  _rippleHighlighted = rippleHighlighted;
+  // Go into the highlighted state visually.
+  if (rippleHighlighted && !_tapWentInsideOfBounds) {
+    // If ripple becomes highlighted we initiate a ripple with animation.
+    [self updateRippleColor];
+    [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
+  } else {
+    if (!self.allowsSelection && !self.selected && !self.dragged && !_tapWentOutsideOfBounds) {
+      // We dissolve the ripple when highlighted is NO, unless we are going into selection.
+      [self updateRippleColor];
+      [self beginRippleTouchUpAnimated:YES completion:nil];
+    }
+  }
+}
+
+- (void)setDragged:(BOOL)dragged {
+  if (dragged == _dragged) {
+    return;
+  }
+  _dragged = dragged;
+  // Go into the dragged state visually.
+  if (dragged) {
+    if (!self.activeRippleLayer) {
+      // If we go into the dragged state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer dragging, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  UITouch *touch = [touches anyObject];
+  CGPoint location = [touch locationInView:self];
+  _didReceiveTouch = YES;
+  _lastTouch = location;
+  _tapWentInsideOfBounds = NO;
+  _tapWentOutsideOfBounds = NO;
+  [super touchesBegan:touches withEvent:event];
+}
+
+- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  // When the touch is held and moved outside and inside the bounds of the surface,
+  // the ripple should gracefully fade out and in accordingly.
+  UITouch *touch = [touches anyObject];
+  CGPoint location = [touch locationInView:self];
+  BOOL pointContainedinBounds = CGRectContainsPoint(self.bounds, location);
+  if (pointContainedinBounds && _tapWentOutsideOfBounds) {
+    if (_tapWentOutsideOfBounds) {
+      _tapWentInsideOfBounds = YES;
+      _tapWentOutsideOfBounds = NO;
+    }
+    [self fadeInRippleAnimated:YES completion:nil];
+  } else if (!pointContainedinBounds && !_tapWentOutsideOfBounds) {
+    _tapWentOutsideOfBounds = YES;
+    [self fadeOutRippleAnimated:YES completion:nil];
+  }
+  [super touchesMoved:touches withEvent:event];
+}
+
+- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  [super touchesEnded:touches withEvent:event];
+  _didReceiveTouch = NO;
+  if (_tapWentOutsideOfBounds) {
+    [self beginRippleTouchUpAnimated:NO completion:nil];
+  }
+}
+
+- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  [super touchesCancelled:touches withEvent:event];
+  _didReceiveTouch = NO;
```

Comment:
I don’t see any reads of this value, consider deleting it.

Reply:
It is read in line 167 and lets us know if the highlighted state was triggered from a touch or manually, and therefore either triggers a ripple animation or not.

Reply:
Got it - thank you for clarifying!